source/swap/contracts/Swap.sol

```diff
@@ -361,17 +362,18 @@ contract Swap is ISwap {
 
     // Ensure the transfer is not to self.
     require(from != to, "INVALID_SELF_TRANSFER");
-
-    if (kind == ERC721_INTERFACE_ID) {
-
-      require(amount == 0, "NO_AMOUNT_FIELD_IN_ERC721");
-      // Attempt to transfer an ERC-721 token.
-      IERC721(token).transferFrom(from, to, id);
-
-    } else {
-      require(id == 0, "NO_ID_FIELD_IN_ERC20");
-      // Attempt to transfer an ERC-20 token, underlying SafeERC20 calls require.
-      IERC20(token).safeTransferFrom(from, to, amount);
-    }
+    ITransferHandler transferHandler = registry.getTransferHandler(kind);
+    require(address(transferHandler) != address(0), "UNKNOWN_TRANSFER_HANDLER");
+    // delegatecall required to pass msg.sender as Swap contract
```

Comment:
This comment is incorrect - the msg.sender isn't being passed as the Swap contract

Reply:
I can reword. The transferTokens -> (token).safeTransferFrom is called and works because the Swap contracts is approved to transfer on behalf of the counter parties. The call to the handler occurs in the context of the calling contract and msg.sender and msg.value remain the same.

Reply:
I see, you mean to pass the msg.sender as swap contract to the token itself. That makes sense.